include/solvers/linear_solver.h

```diff
@@ -257,8 +265,21 @@ class LinearSolver : public ReferenceCountedObject<LinearSolver<T>>,
    */
   void set_solver_configuration(SolverConfiguration & solver_configuration);
 
-protected:
+  /**
+   * Attach a system. This will error if if a system has already been attached
+   */
+  void attach_system(System & sys);
+
+  /**
+   * @returns the attached system
+   */
+  System & system();
 
+protected:
+  /**
+   * The system this solver is solving
+   */
+  System * _system;
```

Comment:
So ... I still hate this and want to get rid of it.

And I'm not seeing why we can't get rid of it by moving it to `PetscLinearSolver`.  If there's nothing at the `LinearSolver` level that uses `_system`, the `LinearSolver` doesn't need to be the one to keep track of `_system`, does it?  Could we handle this more like `init_names(System &)`, where user classes provide the System access later when the solver is potentially able to use it for advanced features?

It just feels like wrong OOP.  If you find yourself writing `Shape::set_radius()` because it was useful for `Circle` then something's off.

Reply:
(But to be fair, if I haven't at least changed @jwpeterson's mind with that, we can just mark this resolved, and I'll consider myself outvoted)